web/web.go

```diff
@@ -189,6 +223,7 @@ func New(logger log.Logger, o *Options) *Handler {
 		},
 		o.Flags,
 		h.testReady,
+		instrf,
```

Comment:
Passing a metric down to a dependency is a bit of a smell. Could we do this in the router rather than on each of the routes individually?

Reply:
Agree that passing the instrumentation handler to build the API isn't required. I'm not sure what you mean by "could we do this in the router".
What I have in mind is to pass the instrumentation handler to the `Register()` method since it is the only one needing the handler.

https://github.com/prometheus/prometheus/blob/7b7e3506acb5e0e5f19de26d89bf9029d6e86ddf/web/api/v1/api.go#L155-L157

to

```
func (api *API) Register(r *route.Router, instrh func(handlerName string, handler http.HandlerFunc) http.HandlerFunc) {
...
```

Reply:
I'd go for the route that is passed in to already have everything needed

Reply:
I opened https://github.com/prometheus/common/pull/120. Is that what you had in mind?